staging/src/k8s.io/apiserver/pkg/authentication/request/anonymous/anonymous.go

```diff
@@ -17,27 +17,43 @@ limitations under the License.
 package anonymous
 
 import (
+	"fmt"
 	"net/http"
 
+	"k8s.io/apiserver/pkg/apis/apiserver"
 	"k8s.io/apiserver/pkg/authentication/authenticator"
 	"k8s.io/apiserver/pkg/authentication/user"
 )
 
 const (
-	anonymousUser = user.Anonymous
-
+	anonymousUser        = user.Anonymous
 	unauthenticatedGroup = user.AllUnauthenticated
 )
 
-func NewAuthenticator() authenticator.Request {
-	return authenticator.RequestFunc(func(req *http.Request) (*authenticator.Response, bool, error) {
-		auds, _ := authenticator.AudiencesFrom(req.Context())
-		return &authenticator.Response{
-			User: &user.DefaultInfo{
-				Name:   anonymousUser,
-				Groups: []string{unauthenticatedGroup},
-			},
-			Audiences: auds,
-		}, true, nil
-	})
+type Authenticator struct {
+	allowedPaths map[string]bool
+}
+
+func (a *Authenticator) AuthenticateRequest(req *http.Request) (*authenticator.Response, bool, error) {
+	if len(a.allowedPaths) > 0 && !a.allowedPaths[req.URL.Path] {
+		return nil, false, fmt.Errorf("anonymous auth is not supported for path %q", req.URL.Path)
```

Comment:
should this return `nil, false, nil`, so the authenticator is not even active for this path (rather than erroring)?

can you check what the difference in behavior would be (in the apiserver log and in the response to the user)?

Reply:
With the error kube-apiserver logs reports
```
E0627 17:22:01.733808       1 authentication.go:73] "Unable to authenticate the request" err="anonymous auth is not supported for path \"/api/v1/pods\""
```

and the user sees

```
curl -k https://127.0.0.1:43617/api/v1/pods
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}%       
```

Without the error, i.e. switching to `nil, false, nil` we don't get any error logs in kube-apiserver and the out put to the user is the same

```
curl -k https://127.0.0.1:41455/api/v1/pods
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}%                                     
```

I think making it `nil, false, nil` make sense as the error log might mislead someone into thinking that there is misconfiguration in the authentication config causing the error.

Reply:
I agree... logging an error because the admin's configuration was effective doesn't seem right